src/mem/metaslab.h

```diff
@@ -1,20 +1,189 @@
+    MetaEntryBase::REMOTE_BACKEND_MARKER;
+  static_assert(
+    (REMOTE_MIN_ALIGN >> 1) == MetaEntryBase::REMOTE_BACKEND_MARKER);
+
+  /**
+   * Entry stored in the pagemap.  See docs/AddressSpace.md for the full
+   * MetaEntry lifecycle.
+   */
+  template<typename BackendMetadata>
+  class MetaEntry : public MetaEntryBase
+  {
+    template<typename Pagemap>
+    friend class BuddyChunkRep;
+
+    /**
+     * In common cases, the pointer to the metaslab.  See docs/AddressSpace.md
+     * for additional details.
+     *
+     * The bottom bit is used to indicate if this is the first chunk in a PAL
+     * allocation, that cannot be combined with the preceeding chunk.
+     */
+    uintptr_t meta{0};
+
+    /**
+     * In common cases, a bit-packed pointer to the owning allocator (if any),
+     * and the sizeclass of this chunk.  See mem/metaslab.h:MetaEntryRemote for
+     * details of this case and docs/AddressSpace.md for further details.
+     */
+    uintptr_t remote_and_sizeclass{0};
+
+  public:
+    constexpr MetaEntry() = default;
+
+    /**
+     * Constructor, provides the remote and sizeclass embedded in a single
+     * pointer-sized word.  This format is not guaranteed to be stable and so
+     * the second argument of this must always be the return value from
+     * `get_remote_and_sizeclass`.
+     */
+    SNMALLOC_FAST_PATH
+    MetaEntry(BackendMetadata* meta, uintptr_t remote_and_sizeclass)
+    : meta(unsafe_to_uintptr<BackendMetadata>(meta)),
+      remote_and_sizeclass(remote_and_sizeclass)
+    {}
+
+    /**
+     * Encode the remote and the sizeclass.
+     */
+    [[nodiscard]] static SNMALLOC_FAST_PATH uintptr_t
+    encode(RemoteAllocator* remote, sizeclass_t sizeclass)
+    {
+      /* remote might be nullptr; cast to uintptr_t before offsetting */
+      return pointer_offset(
+        reinterpret_cast<uintptr_t>(remote), sizeclass.raw());
+    }
+
+    /**
+     * Return the remote and sizeclass in an implementation-defined encoding.
+     * This is not guaranteed to be stable across snmalloc releases and so the
+     * only safe use for this is to pass it to the two-argument constructor of
+     * this class.
+     */
+    [[nodiscard]] SNMALLOC_FAST_PATH const uintptr_t&
+    get_remote_and_sizeclass() const
+    {
+      return remote_and_sizeclass;
+    }
+
+    MetaEntry(const MetaEntry&) = delete;
+
+    MetaEntry& operator=(const MetaEntry& other)
+    {
+      // Don't overwrite the boundary bit with the other's
+      meta = (other.meta & ~META_BOUNDARY_BIT) |
+        address_cast(meta & META_BOUNDARY_BIT);
+      remote_and_sizeclass = other.remote_and_sizeclass;
+      return *this;
+    }
+
+    /**
+     * Return the Metaslab metadata associated with this chunk, guarded by an
+     * assert that this chunk is being used as a slab (i.e., has an associated
+     * owning allocator).
+     */
+    [[nodiscard]] SNMALLOC_FAST_PATH BackendMetadata* get_metaslab() const
+    {
```

Comment:
The assert mentioned in the documentation here has gone missing.  `meta` only encodes a `BackendMetadata*` if there's an associated `RemoteAllocator*` (which should be equivalent to `MetaEntryBase::REMOTE_BACKEND_MARKER` being clear in `remote_and_sizeclass`).

```suggestion
    {
      SNMALLOC_ASSERT(get_remote(me) != nullptr);
```

Reply:
Ooops, this got lost in the merging of `get_metadata` and `get_meta`.